app/APKServer.m

```diff
@@ -0,0 +1,139 @@
+    };
+    NSData *sin_data = [NSData dataWithBytes:&sin length:sizeof(sin)];
+    if (CFSocketSetAddress(socket, (CFDataRef) sin_data) != kCFSocketSuccess) {
+        NSLog(@"APK server: failed to bind");
+        return;
+    }
+    self.server = [[NSFileHandle alloc] initWithFileDescriptor:CFSocketGetNative(socket) closeOnDealloc:YES];
+    [NSNotificationCenter.defaultCenter addObserver:self selector:@selector(onAccept:) name:NSFileHandleConnectionAcceptedNotification object:self.server];
+    [self.server acceptConnectionInBackgroundAndNotify];
+    NSLog(@"APK server: serving");
+}
+
+- (void)onAccept:(NSNotification *)n {
+    NSFileHandle *client = n.userInfo[NSFileHandleNotificationFileHandleItem];
+    CFDictionaryAddValue(self.messages, (__bridge CFTypeRef) client, CFHTTPMessageCreateEmpty(kCFAllocatorDefault, YES));
+    [NSNotificationCenter.defaultCenter addObserver:self selector:@selector(onReadable:) name:NSFileHandleDataAvailableNotification object:client];
+    [client waitForDataInBackgroundAndNotify];
+    [self.server acceptConnectionInBackgroundAndNotify];
+}
+
+- (void)onReadable:(NSNotification *)n {
+    NSFileHandle *client = n.object;
+    CFHTTPMessageRef message = (CFHTTPMessageRef) CFDictionaryGetValue(self.messages, (__bridge CFTypeRef) client);
+    NSData *data = [client availableData];
+    if (data.length == 0)
+        return [self endRequest:client];
+    if (!CFHTTPMessageAppendBytes(message, data.bytes, data.length))
+        return [self endRequest:client];
+    if (CFHTTPMessageIsHeaderComplete(message))
+        return [self sendResponseForRequest:message toClient:client];
+    [client waitForDataInBackgroundAndNotify];
+}
+
+- (void)sendResponseForRequest:(CFHTTPMessageRef)message toClient:(NSFileHandle *)client {
+    NSURL *url = CFBridgingRelease(CFHTTPMessageCopyRequestURL(message));
+    NSArray<NSString *> *components = url.pathComponents;
+    if ([components.firstObject isEqualToString:@"/"])
+        components = [components subarrayWithRange:NSMakeRange(1, components.count - 1)];
+    NSString *resourceTag = [components componentsJoinedByString:@":"];
+
+    NSLog(@"APK server: fetching %@", resourceTag);
+    NSBundleResourceRequest *request = [[NSBundleResourceRequest alloc] initWithTags:[NSSet setWithObject:resourceTag]];
+    request.loadingPriority = NSBundleResourceRequestLoadingPriorityUrgent;
+    [request beginAccessingResourcesWithCompletionHandler:^(NSError * _Nullable error) {
+        int status = 200;
+        NSData *body = nil;
+        NSFileHandle *bodyFile = nil;
+        NSUInteger contentLength = 0;
+
+        if (error == nil) {
+            NSLog(@"APK server: serving %@", resourceTag);
+            bodyFile = [NSFileHandle fileHandleForReadingFromURL:[request.bundle URLForResource:resourceTag withExtension:nil] error:&error];
+            [bodyFile seekToEndOfFile];
+            contentLength = bodyFile.offsetInFile;
+            [bodyFile seekToFileOffset:0];
```

Comment:
Is there no better way to find the file size than this?

Reply:
NSFileManager is surprisingly bad at this